Bb chapter 5 (#697)
mlr-org/mlr3book: book/chapters/chapter5/advanced_tuning_methods_and_black_box_optimization.qmd

)
```

We now construct the tuning instance and a hyperband tuner with `eta = 2`.

> We use `trm("none")` and set the `repetitions` control parameter to `1`, causing Hyperband to terminates itself after all brackets have been evaluated a single time.

Reviewer: ```suggestion
We use `trm("none")` and set the `repetitions` control parameter to `1` so that Hyperband can terminate itself after all brackets have been evaluated a single time.
```

The design of BO is more complex than what we have seen so far in other tuning methods so to help motivate this we will spend a little more time in this section on theory and methodology.

In `r index('hyperparameter optimization')` (@sec-optimization), learners are passed a hyperparameter configuration and evaluated on a given task via a resampling technique to estimate its generalization performance with the goal to find the optimal hyperparameter configuration.
In general, no analytical description for the mapping from hyperparameter configuration to performance exists and gradient information is also not available.
HPO is, therefore, a prime example for `r index('black-box optimization', aside = TRUE)`, which considers the optimization of a function whose mathematical structure and analytical description is unknown or unexploitable.
As a result, the only observable information is the output value (i.e., generalization performance) of the function given an input value (i.e., hyperparameter configuration).
In fact, as evaluating the performance of a learner can take a substantial amount of time, HPO is quite an expensive black-box optimization problem.

> In the real world, black-box optimization problems are encountered quite often in engineering, for example, modeling experiments like crash tests or chemical reactions.

Reviewer: ```suggestion
Black-box optimization problems occur in the real-world, for example they are encountered quite often in engineering such as in modeling experiments like crash tests or chemical reactions.
```

### Building Blocks of Bayesian Optimization {#sec-bayesian-optimization-blocks}

> Bayesian optimization (BO) is a global optimization algorithm that usually follows the following process, displayed in @fig-optimization-loop:

Reviewer: ```suggestion
Bayesian optimization (BO) is a global optimization algorithm that usually follows the following process (@fig-optimization-loop):
```

  geom_line(aes(x = x, y = y), data = xydt) +
  geom_point(aes(x = x, y = y), colour = "black", fill = "black", size = 2, pch = 24,
    data = instance$archive$data[batch_nr == 1]) +
  geom_point(aes(x = x, y = y, fill = batch_nr), size = 2, pch = 21, color = "black",
    data = instance$archive$data[batch_nr > 1]) +
  scale_fill_gradient(low = "black", high = "white") +
  labs(colour = "Evaluation Number") +
  theme_minimal() +
  theme(legend.position = "bottom")
```

> 

Reviewer: ```suggestion
<!-- fixme: rs check plot output with fig-alt -->
```

  scale_colour_manual(values = c("black", "darkgrey")) +
  scale_fill_manual(values = c("black", "darkgrey")) +
  labs(y = "Best Observed Function Value", x = "Number of Function Evaluations", colour = "Method", fill = "Method", linetype = "Method") +
  theme_minimal() +
  theme(legend.position = "bottom")
ggsave("Figures/bo_1d_sinusoidal_bo_rs.pdf", plot = g, width = 6, height = 4)
ggsave("Figures/bo_1d_sinusoidal_bo_rs.png", plot = g, width = 6, height = 4)
```

```{r, echo=FALSE}
#| echo: false
#| label: fig-bayesian-sinusoidal_bo_rs
#| fig-cap: Anytime performance of BO and random search on the 1D sinusoidal function given a budget of 20 function evaluations. Solid line depicts the best observed target value averaged over 10 replications. Ribbons represent standard errors.
#| fig-alt: Anytime performance of BO and random search on the 1D sinusoidal function given a budget of 20 function evaluations. Both perform similarly for the initial design size but BO is much more sample efficient than random search and quickly converges to the global optimum after around 15 function evaluations.
knitr::include_graphics("Figures/bo_1d_sinusoidal_bo_rs.png")
```

> 

Reviewer: ```suggestion
<!-- fixme: rs check plot output with fig-alt -->
```

#| label: fig-pareto-bayesopt
#| fig-cap: Pareto front of TPR and FPR obtained via ParEGO. White dots represent tested configurations, each black dot individually represents a Pareto-optimal configuration and all blue dots together represent the Pareto front.
#| fig-alt: Scatter plot with classif.tpr on x-axis (between 0.75 and 1.00) and classif.fpr on y-axis (between 0.2 and 1.0). The Pareto front is shown as the set of points at roughly (0.85, 0.23), (0.88, 0.24), (0.91, 0.25), (0.93, 0.3), (0.95, 0.35), (0.96, 0.4), (1.00, 0.8).

> ggplot(aes(y = classif.tpr, x = classif.fpr),

Reviewer: ```suggestion
<!-- fixme: rs check plot output with fig-alt -->
ggplot(aes(y = classif.tpr, x = classif.fpr),
```

## Conclusion

In this chapter, we looked at advanced tuning methods.
We started by thinking about the types of errors that can occur during tuning and how to handle these to ensure your HPO process does not crash.
We presented multi-objective tuning, which can be used to optimize performance measures simultaneously.
We then looked at multi-fidelity tuning, in which the Hyberband tuner can be used to efficiently tune algorithms by making use of lower-fidelity evaluations to approximate full-fidelity model performance.
We will return to Hyperband in @sec-hyperband-example-svm where we will learn how to make use of pipelines in order to tune any algorithm with Hyperband.
Finally, we took a deep dive into Bayesian optimization to look at how `r bbotk`, `r mlr3mbo`, and `r mlr3tuning` can be used together to implement complex BO tuning algorithms in `mlr3`, allowing for highly flexible and sample-efficient algorithms.
In the next chapter we will look at feature selection and see how `r mlr3filters` and `r mlr3fselect` use a very similar design interface to `mlr3tuning`.

@tbl-api-advanced-tuning summarizes the most important classes, functions, and methods seen in this chapter.

> 

Reviewer: ```suggestion
@tbl-api-advanced-tuning summarizes the most important classes, functions, and methods seen in this chapter.

```

| `r ref("TuningInstanceMultiCrit")` | `r ref("ti()")`/`r ref("tune()")` | `$result`; `$archive` |
| `r ref("mlr3hyperband::TunerHyperband")` | `tnr("hyperband")`| - |
| `r ref("bbotk::Objective")` | - |  |
| `r ref("bbotk::OptimInstanceSingleCrit")` or `r ref("bbotk::OptimInstanceMultiCrit")` | `r ref("bb_optimize()")` | `$result`; `$archive` |
| `r ref("SurrogateLearner")` | `r ref("srlrn()")`|  |
| `r ref("AcqFunction")` | `r ref("acqf()")` |  |
| `r ref("AcqOptimizer")` | `r ref("acqo()")` |  |
| - | `r ref("loop_function")` | - |
| `r ref('OptimizerMbo')` | `bbotk::opt("mbo")` | |
| `r ref('TunerMbo')` | `tnr("mbo")` | |
| `r ref('Design')` | `r ref("generate_design_random")`; `r ref("generate_design_grid")`; `r ref("generate_design_lhs")`; `r ref("generate_design_sobol")`;| `$data` |

: Important classes and functions covered in this chapter with underlying class (if applicable), class constructor or function, and important class methods (if applicable). {#tbl-api-advanced-tuning}

> 

Reviewer: ```suggestion
<!-- FIXME: rs check output -->
```

## Conclusion

In this chapter, we looked at advanced tuning methods.
We started by thinking about the types of errors that can occur during tuning and how to handle these to ensure your HPO process does not crash.
We presented multi-objective tuning, which can be used to optimize performance measures simultaneously.
We then looked at multi-fidelity tuning, in which the Hyberband tuner can be used to efficiently tune algorithms by making use of lower-fidelity evaluations to approximate full-fidelity model performance.
We will return to Hyperband in @sec-hyperband-example-svm where we will learn how to make use of pipelines in order to tune any algorithm with Hyperband.
Finally, we took a deep dive into Bayesian optimization to look at how `r bbotk`, `r mlr3mbo`, and `r mlr3tuning` can be used together to implement complex BO tuning algorithms in `mlr3`, allowing for highly flexible and sample-efficient algorithms.

> In the next chapter we will look at feature selection and see how `r mlr3filters` and `r mlr3fselect` use a very similar design interface to `mlr3tuning`.

Reviewer: ```suggestion
In the next chapter we will look at feature selection and see how `r mlr3filters` and `r mlr3fselect` use a very similar design interface to `mlr3tuning`.

@tbl-api-advanced-tuning summarizes the most important classes, functions, and methods seen in this chapter.
```
